File: leetcode/LEET21.java
public class LEET21 {
    
    //   Definition for singly-linked list.
    public static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
    
    public static void main(String[] args){
        ListNode node1= new ListNode(1);
        ListNode node2= new ListNode(2);
        ListNode node3= new ListNode(4);

        node1.next = node2;
        node2.next = node3;

        ListNode node4= new ListNode(1);
        ListNode node5= new ListNode(3);
        ListNode node6= new ListNode(4);

        node4.next = node5;
        node5.next = node6;


        System.out.println(mergeTwoLists(node1, node4).value);
    }
    public static ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode root = new ListNode(0);
        ListNode current = root;
        int carry = 0;

        while(l1 != null || l2 != null || carry != 0){
            int sum = carry;
            if(l1 != null){
                sum += l1.val;
                l1 = l1.next;
            }

            if (l2 != null){
                sum += l2.val;
                l2 = l2.next;
            }

            carry = sum / 10;
            int value = sum % 10;

            current.next = new ListNode(value);
            current = current.next;
        }

        return root.next;
    }

}

File: leetcode/python/linked_list/LEET21.py
from typing import *

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def mergeTwoLists(l1: ListNode, l2: ListNode) -> ListNode:
    root = current = ListNode()

    while l1 or l2:
        value = 0
        if l1 and l2:
            if l1.val <= l2.val:
                value = l1.val
                l1 = l1.next
            else:
                value = l2.val
                l2 = l2.next
        elif l1:
            value = l1.val
            l1 = l1.next
        else:
            value = l2.val
            l2 = l2.next
        
        current.next = ListNode(value)
        current = current.next
    return root.next
def mergeTwoLists2(l1: ListNode, l2: ListNode) -> ListNode:
    if not l1 or (l2 and l1.val > l2.val):
        l1, l2 = l2, l1
    
    if l1:
        l1.next = mergeTwoLists2(l1.next, l2)
    
    return l1
def solve():
    n1 = ListNode(1)
    n2 = ListNode(2)
    n3 = ListNode(4)
    n1.next = n2
    n2.next = n3

    m1 = ListNode(1)
    m2 = ListNode(3)
    m3 = ListNode(4)

    m1.next = m2
    m2.next = m3

    result = mergeTwoLists(n1, m1)
    print(result)

solve()